routing.md

```diff
@@ -305,12 +321,51 @@ DIDComm DID Document endpoints have the following format:
 
 **type**: MUST be `DIDComm`. 
 
-**serviceEndpoint**: MUST contain a URI for a transport specified in the [transports] section of this spec.
+**serviceEndpoint**: MUST contain a URI for a transport specified in the [transports] section of this spec, or a URI from Alternative Endpoints.
 
 **routingKeys**: An optional ordered array of strings referencing keys to be used when preparing the message for transmission as specified in the [Routing] section of this spec. 
 
 #### Multiple Endpoints
 
 A DID Document may contain multiple service entries of type `DIDComm`. Unless otherwise specified, messages should be sent to endpoints in the order they are specified.
 
+#### Alternative Endpoints
+
+In addition to the URIs for [transports], some alternative forms are available.
+
+##### DID
+
+Using a DID for the `serviceEndpoint` is useful when using a mediator. The DID should be resolved, and services with type of "DIDComm" will contain valid `serviceEndpoints`. The keyAgreement keys of that DID Document should be appended at the end of the routingKeys section from the message recipient's DID Document as per the process in [Sender Forward Process]. The key advantage with this approach is that a mediator can rotate keys and update serviceEndpoints without any updates needed to dependent recipients` DID Documents.
+
+A DID used as a mediator in this way MUST NOT use alternative endpoints. Only URIs described in [transports] are acceptable.
+
+Example 1: Mediator
+
+```json
+{
+    "id": "did:example:123456789abcdefghi#didcomm-1",
+    "type": "DIDComm",
+    "serviceEndpoint": "did:example:somemediator"
+}
+```
+The message is encrypted to the recipient, then wrapped in a forward message encrypted to the keyAgreement keys within the `did:example:somemediator` DID Document, and transmitted to the URIs present in the `did:example:somemediator` DID Document with type `DIDComm`.
+
+Example 2: Mediator + Routing Keys
+```json
+{
+    "id": "did:example:123456789abcdefghi#didcomm-1",
+    "type": "DIDComm",
+    "serviceEndpoint": "did:example:somemediator",
+    "routingKeys": ["did:example:anothermediator#somekey"]
```

Comment:
As we discussed awhile back, I don't believe this is going to be a usable method of specifying a route. Putting this list of keys in the did document, where the route will likely need to be changing quite frequently (meaning keys will be updated frequently) will almost certainly cause us a large number of headaches. I'd suggest we find a different method other than this. I personally think the JIT routing scheme would be a better approach because we would not need to put routing keys in the did document.

Reply:
We do need a discussion of this, but I wanted to note something important: The list of routingKeys is NOT the complete list of keys in the route. Any number of additional routing steps can exist in between and after the keys listed here. The keys listed here are ONLY the keys shared to the sender from the recipient. 

Use of these keys allow hiding the header information of the final message from mediators on the route, a feature that I don't believe pure JIT routing (only one mediator in the DID Doc) allows.

Reply:
> 
> 
> We do need a discussion of this, but I wanted to note something important: The list of routingKeys is NOT necessarily the complete list of keys in the route. Any number of additional routing steps can exist in between and after the keys listed here. The keys listed here are ONLY the keys shared to the sender from the recipient.
> 
> Use of these keys allow hiding the header information of the final message from mediators on the route, a feature that I don't believe pure JIT routing (only one mediator in the DID Doc) allows.

Which header information are we hiding? The only thing I can think of is the recipient information. Onion routing doesn't solve this completely either as the last node (referred to as the exit node in TOR) is still able to identify the recipient. I'd like us to really assess what the tradeoffs are here. Using onion routing feels like an extension use case rather than the default requirement and I don't think we should be intentionally designing to meet it's needs. Rather we should design for simplicity and allow the people who want to extend to support this functionality to define how that should be done.

Reply:
The recipient information is precisely what we are trying to hide, from all but the last step.

And yes, we should discuss the tradeoffs. We may not be able to complete that conversation in today's meeting, as the US holiday may not allow attandance of key folks for the conversation.

Reply:
Ok, let's dig a bit deeper on this here while we can. Who MUST we hide the recipient info from and/or who SHOULD we hide the recipient info from? Let's start there to keep the conversation constrained (you know how much I love to squirrel in topics like this).

Reply:
@dhh1128 @tplooker @swcurran you'll probably want to be apart of this conversation.

Reply:
It seems like we are thinking of JIT and onion routing as opposites. That's a useful simplification, sometimes -- but it's a distinction that breaks down under various use cases. Of course we can have the simple case where Bob sends to Alice via her agency and via her cloud mediator. We've discussed that at length. But there is nothing to prevent far more complex scenarios. Suppose Alice wants to advertise her public mediator to be AgentsRUs's DID, which has separate instructions to forward to her cloud agent, which which does not appear in her `routingKeys` array at all. In such a case, the cloud agent is incapable of reading a message (even a forward) -- but perhaps it is configured as a relay, which in turn re-encrypts (NOT proxy re-encrypts) and sends the message on to a different DID hosted at a different agency. Repeat 10x. Eventually the message ends up at an agent (NOT a cloud agent that's playing the mediator role) that can directly decrypt the message. This is quite a bit like constructing a channel that consists of interlocking SSH tunnels, where parts of the route (the first one, specifically) use onion routing, and other parts use JIT.

In such a scenario, what does the sender (Bob) need to know about Alice's routing? Just her first public mediator. To him it looks like onion routing, but that's not the whole story, and that's the point. We are NOT representing all the hops in a route with this array. We are representing exactly and only as many hops as the sender, Bob, needs to know to make delivery work. Alice can change the other details to suit her preference.

What will not change is that the final mediator (who unwraps the innermost `forward` message) MUST see the recipients, and all other hops MUST NOT see it unless Alice (not Bob) chooses to expose it.

Reply:
What @dhh1128 said so well.  The receiver tells the sender (via the DIDComm service) to do certain wrappings.  What actually happens in the flow is a separate issue other than the messages must be delivered to the endpoint and at some point be decrypted in order.

Reply:
I love everything about the last few comments here. What is needed to capture this? What can I add to the PR to make this mergable?